subsys/bluetooth/mesh/dfw.c

```diff
@@ -0,0 +1,5398 @@
+				  lane_discovery_guard_get(dfw_cfg.lane_discov_guard_intv));
+		return;
+	}
+
+clear:
+	if (fw && fw->lane_count) {
+		LOG_INF("Path discovery succeeded");
+		dfw_state_machine_event(DFW_STATE_MACHINE_EVENT_PATH_DISCOV_SUCCEED, fw,
+					dv->destination);
+	} else {
+		LOG_INF("Path discovery failed");
+		dfw_state_machine_event(DFW_STATE_MACHINE_EVENT_PATH_DISCOV_FAILED, NULL,
+					dv->destination);
+	}
+
+	dfw_discovery_clear(dv);
+}
+
+static int dfw_path_initialize_start(struct bt_mesh_subnet *sub,
+				     const struct bt_mesh_dfw_node *dependent,
+				     uint16_t dst)
+{
+	struct bt_mesh_dfw_subnet *dfw = sub->dfw;
+	struct bt_mesh_dfw_discovery *dv = NULL;
+	size_t total_concurr_cnt = 0;
+
+	if (dst == BT_MESH_ADDR_ALL_NODES ||
+	    dst == BT_MESH_ADDR_RELAYS) {
+		LOG_ERR("Invalid dst addr provided 0x%04x", dst);
+		return -EINVAL;
+	}
+
+	for (int i = 0; i < ARRAY_SIZE(dfw->discovery); i++) {
+		if (dfw->discovery[i].path_origin.addr == BT_MESH_ADDR_UNASSIGNED) {
+			dv = &dfw->discovery[i];
+		} else if (dfw->discovery[i].path_origin.addr == bt_mesh_primary_addr()) {
+			if (dfw->discovery[i].destination == dst) {
+				LOG_ERR("Path initialize already started for 0x%04x", dst);
+				return -EALREADY;
+			}
+
+			total_concurr_cnt++;
+		}
+	}
+
+	/* If the number of executed Directed Forwarding Initialization procedures is equal
+	 * to the Max Concurrent Init state value, then the Directed Forwarding Initialization
+	 * procedure shall fail.
+	 */
+	if (!dv || total_concurr_cnt == dfw->max_concurr_init) {
+		LOG_ERR("Insufficient discovery for path initialize procedure");
+		return -EBUSY;
+	}
+
+	/* A new entry shall be added to the Discovery Table according to Section 3.6.8.6.1 */
+	dv->path_origin.addr = bt_mesh_primary_addr();
+	dv->path_origin.secondary_count = bt_mesh_elem_count() - 1;
+
+	if (dependent) {
+		dv->dependent_origin[0].addr = dependent->addr;
+		dv->dependent_origin[0].secondary_count = dependent->secondary_count;
+	}
+
+	dv->destination = dst;
+
+	dv->interval = dfw_cfg.discov_intv;
+	dv->lifetime = dfw->lifetime;
+	dv->forwarding_number = ++dfw->forwarding_number;
+	dv->metric = 0;
+
+	k_work_init_delayable(&dv->timer, dfw_initialization_discovery_expired);
+
+	k_work_reschedule(&dv->timer, discovery_timeout_get(dv->interval));
+
+	return dfw_send_path_request(dfw, dv);
+}
+
+static void dfw_discovery_expired(struct k_work *work)
+{
+	struct bt_mesh_dfw_discovery *dv = CONTAINER_OF(work, struct bt_mesh_dfw_discovery,
+							timer.work);
+
+	LOG_DBG("Path discovery timer expired");
+
+	dfw_discovery_clear(dv);
+}
+
+static void dfw_path_lifetime_expired(struct k_work *work)
+{
+	struct bt_mesh_dfw_forwarding *fw = CONTAINER_OF(work, struct bt_mesh_dfw_forwarding,
+							 timer.work);
+	struct bt_mesh_dfw_subnet *dfw = dfw_table_get_by_forward(fw);
+
+	LOG_INF("Path lifetime expired");
+
+	dfw->update_id++;
+
+	dfw_forwarding_clear(fw);
+}
+
+static void dfw_path_reply_delay_expired(struct k_work *work)
```

Comment:
Seems very complex function with multirange nested `if`. Could it be split on simple functions?